pkg/ui/workspaces/cluster-ui/src/api/databases/tableMetaUpdateJobApi.ts

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+import moment from "moment-timezone";
+import { useEffect, useMemo } from "react";
+import useSWR from "swr";
+
+import { fetchDataJSON } from "../fetchData";
+
+const TABLE_META_JOB_API = "api/v2/table_metadata/updatejob/";
+
+export enum TableMetadataJobStatus {
+  NOT_RUNNING = "NOT_RUNNING",
+  RUNNING = "RUNNING",
+}
+
+type TableMetaUpdateJobResponse = {
+  current_status: TableMetadataJobStatus;
+  progress: number;
+  last_start_time?: string;
+  last_completed_time: string;
+  last_updated_time: string;
+  data_valid_duration: number;
+  automatic_updates_enabled: boolean;
+};
+
+type TableMetaUpdateJobInfo = {
+  currentStatus: TableMetaUpdateJobResponse["current_status"];
+  progress: number;
+  lastStartTime: moment.Moment;
+  lastCompletedTime: moment.Moment;
+  lastUpdatedTime: moment.Moment;
+  dataValidDuration: moment.Duration;
+  automaticUpdatesEnabled: boolean;
+};
+
+const getTableMetaUpdateJobInfo = async () => {
+  return fetchDataJSON<TableMetaUpdateJobResponse, null>(TABLE_META_JOB_API);
+};
+
+// useTableMetaUpdateJob is a hook that fetches the current status of the table
+// metadata update job and returns the status and other relevant information.
+// The hook polls the API every 3 seconds if the job status is parsed as running.
+export const useTableMetaUpdateJob = () => {
+  const { data, isLoading, mutate } = useSWR(
+    "tableMetaUpdateJob",
+    getTableMetaUpdateJobInfo,
+    {
+      focusThrottleInterval: 10000,
+      refreshInterval: (latest: TableMetaUpdateJobResponse) => {
+        return latest?.current_status === TableMetadataJobStatus.RUNNING
+          ? 3000
+          : 0;
+      },
+      dedupingInterval: 3000,
+    },
+  );
+
+  const formattedResp: TableMetaUpdateJobInfo = useMemo(() => {
+    if (!data) {
+      return null;
+    }
+    return {
+      currentStatus: data.current_status,
+      progress: data.progress,
+      lastStartTime: data.last_start_time
+        ? moment.utc(data.last_start_time)
+        : null,
+      lastCompletedTime: data.last_completed_time
+        ? moment.utc(data.last_completed_time)
+        : null,
+      lastUpdatedTime: data.last_updated_time
+        ? moment.utc(data.last_updated_time)
+        : null,
+      dataValidDuration: moment.duration(
+        data.data_valid_duration * 1e-6,
+        "millisecond",
+      ),
+      automaticUpdatesEnabled: data.automatic_updates_enabled,
+    };
+  }, [data]);
+
+  const dataValidDurationMs =
+    formattedResp?.dataValidDuration?.asMilliseconds();
```

Comment:
can we guarantee that some of these values are present instead of having to put `?` everywhere? Don't we control the value of `formattedResp`?

Reply:
Removed the optional from datValidDuration as it's not necessary. I also just read that this is preventable through the linter with the following ts rule:
```
    "@typescript-eslint/no-unnecessary-condition": "warn",

```

I'll introduce this in a follow-up PR.

Reply:
Nvm that rule does not play nicely with the protobuf generated types. It also needs strict null checks to be on which we have disabled currently.

Reply:
Oh well. Thanks for trying!